### PR TITLE
_id and _lastUpdated search bindvariables are not added #577

### PR DESCRIPTION
**Pull request by prb112**
- Add Bind Variables to Inclusion and Sorted Query Segment Aggregators
- Verify with tests
	- Whole System Search with SORT and without SORT
	- Resource Specific Search with SORT and without SORT

Signed-off-by: Paul Bastide <pbastide@us.ibm.com>

### CODE REVIEW

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/InclusionQuerySegmentAggregator.java [hunk 1]
```diff
@@ -133,17 +133,21 @@ protected SqlQueryData buildCountQuery() throws Exception {
...
         buildFromClause(queryString, resourceType.getSimpleName());
-        allBindVariables.addAll(this.idsObjects);
-        buildWhereClause(queryString, null);
 
+        // An important step here is to add _id and _lastUpdated and then
+        // the regular bind variables. 
+        List<Object> allBindVariables = new ArrayList<>();
+        allBindVariables.addAll(idsObjects);
+        allBindVariables.addAll(lastUpdatedObjects);
+        this.addBindVariables(allBindVariables);
```
Comment (albertwang-ibm) [inline fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/InclusionQuerySegmentAggregator.java]: are the other regular bind variables are added in this.addBindVariables(allBindVariables)?

Comment (albertwang-ibm) [reply fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/InclusionQuerySegmentAggregator.java]: just checked the codes, and found it's

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/InclusionQuerySegmentAggregator.java [hunk 2]
```diff
@@ -232,13 +242,14 @@ private void executeIncludeSubQuery(StringBuilder queryString, InclusionParamete
...
         SqlQueryData subQueryData = new SqlQueryData(subQueryString.toString(), bindVariables);
+
         boolean isFirstItem = true;
         for (String strValue : this.resourceDao.searchStringValues(subQueryData)) {
             if (!isFirstItem) {
                 queryString.append(COMMA);
             }
             if (strValue != null) {
-                queryString.append(QUOTE).append(strValue).append(QUOTE);
+                queryString.append(QUOTE).append(SqlParameterEncoder.encode(strValue)).append(QUOTE);
```
Comment (albertwang-ibm) [inline fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/InclusionQuerySegmentAggregator.java]: what the SqlParameterEncoder.encode is used for here?

File: fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchIdAndLastUpdatedTest.java [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+        try (StringWriter writer = new StringWriter();) {
+            FHIRGenerator.generator(Format.JSON, true).generate(resource, System.out);
+            System.out.println(writer.toString());
+        } catch (FHIRGeneratorException e) {
+            fail("unable to generate the fhir resource to JSON", e);
+        } catch (IOException e1) {
+            fail("unable to generate the fhir resource to JSON (io problem) ", e1);
+        }
+    }
+}
```
Comment (albertwang-ibm) [inline fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchIdAndLastUpdatedTest.java]: Could you add one "gt" test for _lastModified?

Comment (prb112) [reply fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchIdAndLastUpdatedTest.java]: sure

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/InclusionQuerySegmentAggregator.java (hunk 2 above)
Comment (prb112) [reply fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/InclusionQuerySegmentAggregator.java]: avoid secondary injection problems when self assembling a query. https://haiderm.com/second-order-sql-injection-explained-with-example/

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/InclusionQuerySegmentAggregator.java (hunk 1 above)
Comment (prb112) [reply fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/InclusionQuerySegmentAggregator.java]: that's right.

File: fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchIdAndLastUpdatedTest.java (hunk 3 above)
Comment (prb112) [reply fhir-persistence/src/test/java/com/ibm/fhir/persistence/search/test/AbstractSearchIdAndLastUpdatedTest.java]: in next commit

**Pull request merged**